q2_composition/_diff_abundance_plots.py

```diff
@@ -0,0 +1,260 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+from pathlib import Path
+from collections import OrderedDict
+
+import altair as alt
+import pandas as pd
+import numpy as np
+
+import qiime2
+from q2_composition import DataLoafPackageDirFmt
+
+_html_head = """<html>
+<head>
+<style>
+    body {
+        padding: 20px;
+        font-family: Verdana, sans-serif;
+    }
+    div {
+        width: 60%;
+    }
+</style>
+<meta charset="UTF-8">
+</head>
+<body>
+<p>Click a link to see the differential abundance bar plot for the specified
+   category:
+</p>
+<ul>
+"""
+
+_html_foot = """</ul>
+<div>
+<hr>
+<p>Notes on interpreting plots with taxonomic feature identifiers:</p>
+<ul>
+<li>If taxonomic labels are used to identify features, the feature labels
+(y-axis labels) in each plot represent the most specific named taxonomic level
+associated with that feature.</li>
+<li>Hover over the bars in plots to see the full taxonomic label of each
+feature identifier and information about its differential abundance relative
+to the reference.</li>
+<li>Feature identifiers (y-axis labels) that are followed by an asterisk
+(<code>*</code>) represent instances of a duplicated taxonomic name at the
+level displayed in the feature identifier. The number next to the feature
+identifiers in these cases is used only for unique identification in the
+current figure. It is not taxonomically meaningful, and won't be
+consistent across visualizations.</li>
+</ul>
+</div>
+</body>
+</html>
+"""
+
+
+def _plot_differentials(
+        output_dir,
+        df,
+        title,
+        feature_id_label,
+        effect_size_label,
+        significance_label,
+        error_label,
+        feature_ids,
+        effect_size_threshold,
+        significance_threshold):
+
+    if len(df) == 0:
+        raise ValueError("No features present in input.")
+
+    # ensure that all columns are of the types required internally
+    df = df.astype({feature_id_label: pd.StringDtype(),
+                    error_label: pd.Float64Dtype(),
+                    significance_label: pd.Float64Dtype(),
+                    effect_size_label: pd.Float64Dtype()})
+
+    if feature_ids is not None:
+        df = df.query(f'`{feature_id_label}` in @feature_ids')
+
+    df = df[df[significance_label] <= significance_threshold]
+    df = df[np.abs(df[effect_size_label]) >= effect_size_threshold]
+
+    if len(df) == 0:
+        raise ValueError("No features remaining after applying filters.")
+
+    output_dir = Path(output_dir)
+    output_dir.mkdir(parents=True, exist_ok=True)
+    fig_fn = '-'.join(f'{title}-ancombc-barplot.html'.split())
```

Comment:
A better approach would be to use `urllib.parse.quote_plus` on the `title` - this will sanitize basically everything.

As an example of something that would break: consider a column name like "foo/bar"

Reply:
Thanks @ebolyen! I fixed this in my most recent commit.